File: solutions/0739-daily-temperatures/solution.cpp
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        std::vector<int> v_warmer(temperatures.size());
        std::stack<int> m_stack;
        // std::fill(v_warmer.begin(), v_warmer.end(), 0);
        int index = 0;
        for (auto& t: temperatures) {
            while (!m_stack.empty() && temperatures[index] > temperatures[m_stack.top()]) {
                v_warmer[m_stack.top()] = (index - m_stack.top());
                m_stack.pop();
            }
            m_stack.push(index++);
        }
        /* std::cout << "warmer day wait: ";
        for(auto& w: v_warmer) {
            std::cout << "  " << w;
        }
        std::cout << std::endl; */
        return v_warmer;
    }
};

File: solutions/0739-daily-temperatures/solution.go
import lls "github.com/emirpasic/gods/stacks/linkedliststack"

func dailyTemperatures(temperatures []int) []int {
	n := len(temperatures)
	if n == 0 {
		return []int{}
	}
	if n == 1 {
		return []int{temperatures[0]}
	}
	if n == 2 {
		if temperatures[0] > temperatures[1] {
			return []int{temperatures[0]}
		}
		return []int{temperatures[1]}
	}
	warmer := make([]int, n)
	stack := lls.New()
	for i, t := range temperatures {
		var sTop int
        sEle, ok := stack.Peek()
        if ok { sTop = sEle.(int) }
		for ok && t > temperatures[sTop] {
			// Next warmer day is x days away
			warmer[sTop] = (i - sTop)
			stack.Pop()
			sEle, ok = stack.Peek()
            if ok { sTop = sEle.(int) }
		}
		stack.Push(i)
	}
    return warmer
}

